File: dotnet-desktop-guide/winforms/controls/snippets/how-to-make-thread-safe-calls/cs/FormInvokeSync.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace project;

public partial class FormInvokeSync : Form
{
    public FormInvokeSync()
    {
        InitializeComponent();
    }

    //<Good>
    private void button1_Click(object sender, EventArgs e)
    {
        WriteTextSafe("Writing message #1");
        _ = Task.Run(() => WriteTextSafe("Writing message #2"));
    }

    public void WriteTextSafe(string text)
    {
        if (textBox1.InvokeRequired)
            textBox1.Invoke(() => WriteTextSafe($"{text} (NON-UI THREAD)"));

        else
            textBox1.Text += $"{Environment.NewLine}{text}";
    }
    //</Good>

    //<Bad>
    private void button2_Click(object sender, EventArgs e)
    {
        WriteTextUnsafe("Writing message #1 (UI THREAD)");
        _ = Task.Run(() => WriteTextUnsafe("Writing message #2 (OTHER THREAD)"));
    }

    private void WriteTextUnsafe(string text) =>
        textBox1.Text += $"{Environment.NewLine}{text}";
    //</Bad>
}

File: dotnet-desktop-guide/winforms/controls/snippets/how-to-make-thread-safe-calls/vb/FormInvokeSync.vb
Public Class FormInvokeSync
    '<Good>
    Private Sub Button1_Click(sender As Object, e As EventArgs) Handles Button1.Click

        WriteTextSafe("Writing message #1")
        Task.Run(Sub() WriteTextSafe("Writing message #2"))

    End Sub

    Private Sub WriteTextSafe(text As String)

        If (TextBox1.InvokeRequired) Then

            TextBox1.Invoke(Sub()
                                WriteTextSafe($"{text} (NON-UI THREAD)")
                            End Sub)

        Else
            TextBox1.Text += $"{Environment.NewLine}{text}"
        End If

    End Sub
    '</Good>

    '<Bad>
    Private Sub Button2_Click(sender As Object, e As EventArgs) Handles Button2.Click
        WriteTextUnsafe("Writing message #1 (UI THREAD)")
        Task.Run(Sub() WriteTextUnsafe("Writing message #2 (OTHER THREAD)"))
    End Sub

    Private Sub WriteTextUnsafe(text As String)
        TextBox1.Text += $"{Environment.NewLine}{text}"
    End Sub
    '</Bad>
End Class
